doc/developer/design/20220317_liveness_of_unmaterialized.md

```diff
@@ -0,0 +1,57 @@
+# Supporting liveness of weak reads on "unmaterialized" sources
+
+## Summary
+
+[This design document](https://github.com/MaterializeInc/materialize/pull/11262) documents
+how we want to achieve _strict serializability_ as an option for users querying sources
+in `materialized`. In addition, it mentions that currently, "materialized" sources are _serializable_ and "live".
+
+This document describes how we will add this liveness property to "unmaterialized" sources.
+
+## Goals
+
+Offer an eventually-consistent, live experience for "unmaterialized" sources, to match the semantics
+
+## Non-Goals
+
+Alter the default experience of using `materialized` (yet).
+
+## Description
+
+### Background
+
+This is in contrast to the current default for reads against "materialized" sources,
+which is `serializable` (we endeavor for completely consistent reads as any given timestamp),
+but not neccesarily _current_. We do, however, offer "liveness", which means
+we make sure that we are always making progress through the source's upstream.
+
+This is not a property we currently have for "unmaterialized" sources, which have no
+mechanism for ensuring liveness.
+
+We _do_ have a mechanism called _timestamp bindings_, which associated offsets and timestamps, which ensure _serializability_ for `materialized` sources, which have
+a default _index_ that continuously produces new bindings.
+
+### Solution
+
+Instead of creating a default _index_ on "unmaterialized" source creation, we can instead create a special source _instance_, that produces messages (with offsets),
+for which we store no actual data. This _instance_ will cause _timestamp bindings_ for these offets. When reads (also called "peeks") happen involving these
```

Comment:
Today, ts binding persistence alongside stable upstream data is a Kafka specific, sort of hybrid mode. We'll need to preserve that mode on prem, but in general we should be angling towards all sources always being persisted via persist. Given that, how would this proposed solution apply to eg. postgres cdc sources? Will unmaterialized sources be writing data and ts bindings into persist as reads happen?

I'd also be curious if these ts bindings would be used later if the source is materialized.

Reply:
I think these bindings are generally useful, especially for folks who want to ask "how current is your result as a function of my offsets/LSNs/file_lengths". The thing I always want is a time-varying collection containing the max offsets, so that e.g. I can join my results with that, and get the consistent result that is "results, offsets" showing me something accurate in terms of my inputs.

I'd love to have each source provide a "gauge" of progress (a la reclocking) and always produce a "reclocking topic" as part of `CREATE SOURCE`. This might be outside of this doc, but having such a topic and having it always be "materialized" even if the source is not, would I think be illuminating?

Reply:
updated the doc!